File: saveScripts/saveScoreboard.js
var request = require('request');
var fs = require('fs');
var loc = require('./saveLocation');

fs.readFile(loc + 'downloads/round.json', function (err, data) {
	//console.log("james", JSON.parse(data),"James");
	var round = JSON.parse(data).round;
	fs.readFile(loc + 'downloads/currentseason.json', function (err, cS) {
		var currSeas = JSON.parse(cS);
		var roundDates = [];
		for (var i = 0; i < currSeas.length; i++) {
			if (currSeas[i].round === round) {
				roundDates.push(new Date(convertDate(currSeas[i].datetime)));
			}
		}
		var scoreArr = [];
		var uRoundDates = [];
		for (var i = 0; i < roundDates.length; i++) {
			var found = false;
			for (var j = 0; j < uRoundDates.length; j++) {
				if (roundDates[i].getDate() === uRoundDates[j].getDate()) {
					found = true;
				}
			}
			if (!found) {
				uRoundDates.push(roundDates[i]);
			}
		}
		function addZero(num) {
			if (num < 10) {
				return '0' + String(num)
			} else {
				return String(num)
			}
		}
		// Drop Invalid Date entries (e.g. single-digit days like "8/05/2026" that
		// convertDate doesn't zero-pad — fixed below, but guard here too)
		uRoundDates = uRoundDates.filter(function(d) { return !isNaN(d.getTime()); });

		if (uRoundDates.length === 0) {
			fs.writeFile(loc + 'downloads/scoreboard' + round + '.json', JSON.stringify(scoreArr), function (err) {
				console.log('Saved! (no valid dates for round ' + round + ')');
			});
			return;
		}

		var returnTally = 0;
		for (var i = 0; i < uRoundDates.length; i++) {
			var dateTag = String(uRoundDates[i].getFullYear()) + addZero((uRoundDates[i].getMonth() + 1)) + addZero(uRoundDates[i].getDate());
			request({
				gzip: true,
				json: true,
				url: 'https://site.web.api.espn.com/apis/site/v2/sports/rugby/scorepanel?contentorigin=espn&dates=' + dateTag + '&lang=en&region=us'
			}, async function (error, response, body) {
				if (!error && body && body.scores) {
					var legs = body.scores;
					var leg = {};
					for (var i = 0; i < legs.length; i++) {
						if (legs[i].leagues && legs[i].leagues[0] && legs[i].leagues[0].name === "Super Rugby Pacific") {
							leg = legs[i];
						}
					}
					var comps = leg.events;
					if (comps) {
						for (var i = 0; i < comps.length; i++) {
							scoreArr.push(comps[i].competitions[0].competitors[0])
							scoreArr.push(comps[i].competitions[0].competitors[1])
						}
					}
				} else {
					console.log('Scoreboard API error or empty response for date', dateTag, error || (body && body.message) || '');
				}
				returnTally++;
				if (returnTally === uRoundDates.length) {
					console.log(scoreArr)
					fs.writeFile(loc + 'downloads/scoreboard' + round + '.json', JSON.stringify(scoreArr), function (err) {
						console.log('Saved!');
					});
				}
			});
		}

	});
});


function convertDate(date) {
	date = date.split("/");
	var day = ('0' + date[0]).slice(-2);
	var month = ('0' + date[1]).slice(-2);
	var year = date[2].split(" ");
	date = year[0] + "-" + month + "-" + day;
	time = year[1];
	date = date + "T" + time;
	return date;
}

File: scripts/save_scoreboard.py
"""Fetch Super Rugby scores from ESPN for the current round and save them."""
import json
import os
from datetime import datetime, timezone

import requests

ROOT = os.path.join(os.path.dirname(__file__), '..')
DATA_DIR = os.path.join(ROOT, 'data')


def convert_date(date_str):
    parts = date_str.split('/')
    day = parts[0].zfill(2)
    month = parts[1].zfill(2)
    year_time = parts[2].split(' ')
    return f'{year_time[0]}-{month}-{day}T{year_time[1]}'


with open(os.path.join(DATA_DIR, 'round.json')) as f:
    round_no = json.load(f)['round']

with open(os.path.join(DATA_DIR, 'currentseason.json')) as f:
    curr_seas = json.load(f)

round_dates = []
for entry in curr_seas:
    if entry['round'] == round_no:
        try:
            dt = datetime.fromisoformat(convert_date(entry['datetime']))
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            round_dates.append(dt)
        except ValueError:
            pass

# Deduplicate by calendar day
seen_days = set()
unique_dates = []
for d in round_dates:
    key = d.date()
    if key not in seen_days:
        seen_days.add(key)
        unique_dates.append(d)

# Filter out invalid dates
unique_dates = [d for d in unique_dates if d.year > 1]

score_arr = []

if not unique_dates:
    out_path = os.path.join(DATA_DIR, f'scoreboard{round_no}.json')
    with open(out_path, 'w') as f:
        json.dump(score_arr, f)
    print(f'Saved! (no valid dates for round {round_no})')
else:
    completed = 0
    for d in unique_dates:
        date_tag = d.strftime('%Y%m%d')
        base_url = (
            'https://site.web.api.espn.com/apis/site/v2/sports/rugby/scorepanel'
            f'?contentorigin=espn&dates={date_tag}&lang=en'
        )
        try:
            leg = {}
            for region in ('au', 'us', 'gb'):
                resp = requests.get(base_url + f'&region={region}', timeout=15)
                body = resp.json()
                for item in body.get('scores', []):
                    leagues = item.get('leagues', [])
                    if leagues and leagues[0].get('name') == 'Super Rugby Pacific':
                        leg = item
                        break
                if leg:
                    break
            comps = leg.get('events') or []
            for comp in comps:
                competitors = comp['competitions'][0]['competitors']
                score_arr.append(competitors[0])
                score_arr.append(competitors[1])
        except Exception as e:
            print(f'Scoreboard API error for {date_tag}: {e}')
        completed += 1

    print(score_arr)
    out_path = os.path.join(DATA_DIR, f'scoreboard{round_no}.json')
    with open(out_path, 'w') as f:
        json.dump(score_arr, f)
    print('Saved!')
